### DISCUSSION TITLE
Ulauncher window positioning issues (Wayland)

### DISCUSSION THREAD

**Discussion by friday** [Ulauncher/Ulauncher#1350] [Troubleshooting]
The Wayland protocol maintainers and Gnome Desktop have taken a stance against applications being able to position their windows. They think applications should not have this ability. For most apps this makes sense, but Ulauncher is not a regular application.

Modern Wayland desktops except Gnome's window manager Mutter (and the "Muffin" fork of it) have chosen to support this feature anyway, via [wlr layer shell protocol](https://wayland.app/protocols/wlr-layer-shell-unstable-v1).

Ulauncher v6 (currently in beta) supports positioning for x11 and wlr-layer-shell, which covers most desktops. Additionally, since [beta21](https://github.com/Ulauncher/Ulauncher/releases/tag/v6.0.0-beta21) we added a  workaround for Ulauncher to work better under Gnome/Mutter as well. This is implemented by covering the full display size with a window with transparent background, and then render the visible Ulauncher window centered inside of this. This workaround has some side effects. For example you can no longer hold your cursor on top of another window and use the scroll wheel to scroll that window while Ulauncher is showing, and [dash-to-dock will autohide](https://github.com/Ulauncher/Ulauncher/issues/1565) when Ulauncher is opened.

Ulauncher version 5.x.x and older uses XWayland by default when you launch Ulauncher with the Desktop files (including using XDG Autostart) or the systemd service. XWayland solves some of the Wayland issues, but potentially introduces new issues that are worse than using pure Wayland. If you have multiple screens it might open Ulauncher on the wrong one, because it can't know which screen is active. With pure Wayland that's the only screen you can open new windows, so this problem doesn't exist there, but for X11 you had to check the cursor position to calculate this, and in XWayland it's not possible to reliably know the cursor position. If you try to get the cursor position it will report the *last known* cursor position (the position last time you had your cursor over an XWayland window). So it's very likely that this position is wrong.

For a table detailing general Wayland compatibility, see https://github.com/Ulauncher/Ulauncher/discussions/1353

**Comment by zackkrida** [comment]
`gsettings set org.gnome.mutter center-new-windows true` is a good solution to center the ulauncher window in 6.0.0.beta5 in Debian 12. Thanks!

**Reply by friday** [reply]
Note that this suggestion has now been removed from the post as it should no longer be needed as of Ulauncher v6-beta21 But it also won't harm.

**Comment by Nakamerica** [comment]
I'm sure the ulauncher team is sick of hearing about the ulauncher window rendering on the wrong display vis a vis the most recent mouse pointer position, but unfortunately setting `gsettings set org.gnome.mutter center-new-windows true` also does not help on ulauncher v5.15.7 running on GNOME Wayland. I have also removed `GDK_BACKEND=x11` from ulauncher.desktop. This is how my ulauncher.desktop file looks at the moment:

```
cat ulauncher.desktop
[Desktop Entry]
Name=Ulauncher
Comment=Application launcher for Linux
GenericName=Launcher
Categories=GNOME;GTK;Utility;
TryExec=/usr/bin/ulauncher
Exec=env /usr/bin/ulauncher --hide-window
Icon=ulauncher
Terminal=false
Type=Application

```
of course, running ulauncher in a terminal window every time I log in to GNOME is an option (and it works at rendering the window on the correct display), but it is a bit inconvenient to do so every time I log in to GNOME. Currently, I am considering forcing Ulauncher to work through Wayland (editing Exec in the .desktop file) and hoping that it works in rendering the window on the correct display every time.

**Reply by Nakamerica** [reply]
```
cat /usr/share/applications/ulauncher.desktop
[Desktop Entry]
Name=Ulauncher
Comment=Application launcher for Linux
GenericName=Launcher
Categories=GNOME;GTK;Utility;
TryExec=/usr/bin/ulauncher
Exec=/usr/bin/ulauncher --hide-window
Icon=ulauncher
Terminal=false
Type=Application
```

removing env from the Exec variable forces ulauncher to run on wayland, resolving the issue.

**Comment by SturmB** [comment]
Will moving to Ulauncher 6 take care of the too-small Preferences window that cannot be resized on CachyOS using KDE Plasma?

![Screenshot_20250603_220526](https://github.com/user-attachments/assets/a7fd71eb-163a-4392-85ba-2f1fb3ebe537)

**Reply by friday** [reply]
Yes

**Reply by SturmB** [reply]
Yep, it sure does... But at what cost? The main launcher now has an ugly frame around it.

![Screenshot_20250604_201746](https://github.com/user-attachments/assets/f169bb85-61cf-4abc-9748-7c2923d12519)

So, I guess it's a matter of choosing whether I want the prettier launcher or an easier-to-use Preferences window. It's a shame I have to make that choice.

**Reply by friday** [reply]
No need to be rude and entitled. Ulauncher is a free and open source app. Everyone who contributed to it did so on a voluntary basis. I personally have put in thousands of hours into Ulauncher and haven't received a cent for my time. And I didn't create any of these issues. No one "created" issues intentionally to make Ulauncher worse and/or make your experience worse for that matter.

This is also a post titled "Wayland positioning issues", so very clearly the wrong place to discuss this.

The issue that you see is a KDE **feature**. I think it only happens with pure Wayland and it happens in both both Ulauncher V5 and V6 because both are drawing the window in the same way without the default OS titlebar and borders. The difference I suspect is that V6 removed a hack that V5 used to have in place to make Ulauncher default to use XWayland. You can read more about it and how to work around it in the Wayland compatibility discussion post (linked to from the top post here).

We have a guideline to prevent rude people from ruining our days and make us burn out and lose motivation to continue to contribute. If you want to continue talking here you must read and follow these terms. Otherwise we will ban you: https://github.com/Ulauncher/Ulauncher/blob/v6/CODE_OF_CONDUCT.md

**Reply by SturmB** [reply]
You are correct. I am sorry for my behavior; it was uncalled for. At the time, I was feeling frustrated because this was just the latest in a long line of problems I was having migrating to Linux, but I know that's no excuse. Again, my apologies.

Thank you for taking the time to explain the differences and offering a way to work around them; I will check that out this afternoon.
